std/assembly/date.ts

```diff
@@ -1,38 +1,249 @@
+      mills;
+  }
+
+  setUTCFullYear(value: i32): void {
+    const ymd = ymdFromEpochDays(i32(this.epochMillis / MILLIS_PER_DAY));
+    const mills = this.epochMillis % MILLIS_PER_DAY;
+    this.epochMillis =
+      i64(daysSinceEpoch(value, ymd.month, ymd.day)) * MILLIS_PER_DAY + mills;
+  }
+
+  toISOString(): string {
+    const ymd = ymdFromEpochDays(i32(this.epochMillis / MILLIS_PER_DAY));
+
+    let yearStr = ymd.year.toString();
+    if (yearStr.length > 4) {
+      yearStr = "+" + yearStr.padStart(6, "0");
+    }
+
+    return (
+      yearStr +
+      "-" +
+      ymd.month.toString().padStart(2, "0") +
+      "-" +
+      ymd.day.toString().padStart(2, "0") +
+      "T" +
+      this.getUTCHours().toString().padStart(2, "0") +
+      ":" +
+      this.getUTCMinutes().toString().padStart(2, "0") +
+      ":" +
+      this.getUTCSeconds().toString().padStart(2, "0") +
+      "." +
+      this.getUTCMilliseconds().toString().padStart(3, "0") +
+      "Z"
+    );
+  }
+}
+
+function epochMillis(
+  year: i32,
+  month: i32,
+  day: i32,
+  hour: i32,
+  minute: i32,
+  second: i32,
+  milliseconds: i32
+): i64 {
+  return (
+    i64(daysSinceEpoch(year, month, day)) * MILLIS_PER_DAY +
+    hour * MILLIS_PER_HOUR +
+    minute * MILLIS_PER_MINUTE +
+    second * MILLIS_PER_SECOND +
+    milliseconds
+  );
+}
+
+function throwIfNotInRange(value: i32, lower: i32, upper: i32): void {
+  if (value < lower || value > upper) throw new RangeError(E_VALUEOUTOFRANGE);
+}
+
+const MILLIS_PER_DAY = 1_000 * 60 * 60 * 24;
+const MILLIS_PER_HOUR = 1_000 * 60 * 60;
+const MILLIS_PER_MINUTE = 1_000 * 60;
+const MILLIS_PER_SECOND = 1_000;
+
+class YMD {
+  year: i32;
+  month: i32;
+  day: i32;
+}
+
+// http://howardhinnant.github.io/date_algorithms.html#is_leap
+function isLeap(y: i32): bool {
+  return y % 4 == 0 && (y % 100 != 0 || y % 400 == 0);
+}
+
+// http://howardhinnant.github.io/date_algorithms.html#last_day_of_month_common_year
+function lastDayOfMonthNonLeapYear(m: i32): i32 {
+  const days = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
+  return days[m - 1];
+}
+
+// http://howardhinnant.github.io/date_algorithms.html#last_day_of_month
+function lastDayOfMonth(y: i32, m: i32): i32 {
+  return m != 2 || !isLeap(y) ? lastDayOfMonthNonLeapYear(m) : 29;
+}
+
+// see: http://howardhinnant.github.io/date_algorithms.html#civil_from_days
+function ymdFromEpochDays(z: i32): YMD {
+  z += 719468;
+  const era = (z >= 0 ? z : z - 146096) / 146097;
+  const doe = z - era * 146097; // [0, 146096]
+  const yoe = (doe - doe / 1460 + doe / 36524 - doe / 146096) / 365; // [0, 399]
+  const y = yoe + era * 400;
+  const doy = doe - (365 * yoe + yoe / 4 - yoe / 100); // [0, 365]
+  const mp = (5 * doy + 2) / 153; // [0, 11]
+  const d = doy - (153 * mp + 2) / 5 + 1; // [1, 31]
+  const m = mp + (mp < 10 ? 3 : -9); // [1, 12]
+  return {
+    year: y + (m <= 2 ? 1 : 0),
+    month: m,
+    day: d,
+  };
+}
```

Comment:
Could you split this to three separate functions:
`yearFromEpochDays`, `monthFromEpochDays` and `dayFromEpochDays`? This will make returning YMD object unnecessary

Reply:
Most of the time you need multiple components (e.g. year and month), so splitting into separate functions would result in repeated calculations.

Is it better to provide use an array to return these values perhaps?

Reply:
I guess better manually inline everything. At leat for now until we have scalar replacement optimization.
@dcodeIO  wdyt?

Reply:
Another option could be to return just the `year` component, and put `month` and `day` into globals. So one still gets all the values, but without an allocation.

Reply:
Rather than special-case year, I've modified the function to be void, returning all three values as globals https://github.com/AssemblyScript/assemblyscript/pull/1768/commits/03e4d8b8de9df68e6be98cef802225f5741034af

Reply:
@ColinEberhardt ok